### PR TITLE
JP-3242: Proper Handling of Timing for Final Slope Computation for only 0th Good Group Ramps

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves [JP-3242](https://jira.stsci.edu/browse/JP-3242)

This PR addresses special handling of timing for ramps that have only good 0th group in a ramp. The addition of this special case only properly computed the variances. The timing division for the final slope computation still used the group time, which is not correct for certain special cases. All timing divisions have been pushed back to the segment calculations, so there is no longer one final divide. There is also the further subset special case for only 0th good group with the use of ZEROFRAME.

These corrections have been made in STCAL, which affects the CI tests in JWST.  The ramp fitting CI tests have been updated.  The STCAL PR https://github.com/spacetelescope/stcal/pull/173 needs to be merged first before this PR.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Started regtest, using stcal/main, at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/754/

**Comment by hbushouse** [comment]
regtest run showed mostly unrelated differences, except for 1 ramp_fit unit test that's also failing in the CI runs. "No module named 'general_funcs'"

File: jwst/ramp_fitting/tests/test_ramp_fit_cases.py [hunk 1]
```diff
@@ -5,6 +5,18 @@
 
 from stdatamodels.jwst.datamodels import RampModel, GainModel, ReadnoiseModel, dqflags
 
+
+################## DEBUG ################## 
+#                  HELP!!
+import sys
+sys.path.insert(1, "/Users/kmacdonald/code/common")
```
Comment (hbushouse) [inline jwst/ramp_fitting/tests/test_ramp_fit_cases.py:12]: I don't think we want to have this included in the public code base. It's also causing CI and regtest errors.

Comment (kmacdonald-stsci) [reply jwst/ramp_fitting/tests/test_ramp_fit_cases.py]: I removed this.

File: jwst/ramp_fitting/tests/test_ramp_fit.py [hunk 2]
```diff
@@ -744,10 +744,11 @@ def test_twenty_groups_two_segments():
     np.testing.assert_allclose(oslope[wh_data], 10. / deltatime, rtol=1E-4)
     np.testing.assert_allclose(oyint[0, 0, 0, :], model1.data[0, 0, 0, :], rtol=1E-5)
 
-    np.testing.assert_allclose(
-        opedestal[0, 0, :],
-        model1.data[0, 0, 0, :] - 10.,
-        rtol=1E-5)
+    opedal = opedestal[0, 0, :]
```
Comment (hbushouse) [inline jwst/ramp_fitting/tests/test_ramp_fit.py:747]: code style check claims that `opedal` is set but never used

**Comment by hbushouse** [comment]
Looks to me like all the CI failures are due to the use of stcal==1.3.8, so it isn't picking changes from the PR recently merged to stcal/main. Guess we need an stcal release to fix that situation. Although a regtest run making use of stcal/main should get around that.

**Comment by hbushouse** [comment]
Another regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/757 using stcal/main.

File: jwst/ramp_fitting/tests/test_ramp_fit.py (hunk 2 above)
Comment (kmacdonald-stsci) [reply jwst/ramp_fitting/tests/test_ramp_fit.py:747]: Fixed.

**Comment by kmacdonald-stsci** [comment]
> Looks to me like all the CI failures are due to the use of stcal==1.3.8, so it isn't picking changes from the PR recently merged to stcal/main. Guess we need an stcal release to fix that situation. Although a regtest run making use of stcal/main should get around that.

I ran the ramp fitting tests locally using the latest STCAL main and they passed.

**Comment by hbushouse** [comment]
Still has a code style failure for the opedal variable in the unit tests.

**Comment by hbushouse** [comment]
Latest regtest run shows only a few unrelated differences. Looks good.

File: CHANGES.rst [hunk 3]
```diff
@@ -138,6 +138,12 @@ ramp_fitting
   of groups in a segment are computed when applying optimal weighting to line
   fit segments. [#7560, spacetelescope/stcal#163]
 
+- Updated the CI tests due to change in STCAL, which fixed bug for using the
```
Comment (hbushouse) [inline CHANGES.rst:141]: Because this didn't make it into the 1.11.0 release, this entry needs to be moved up to the top of the log.

**Review by hbushouse — APPROVED** [review]
Everything looks good now, except of course for the CI tests that won't pass because they're using the released version of stcal, rather than stcal/main. All tests pass in the regtest environment when build with stcal/main.

**Pull request merged**